cap-primitives/src/rustix/linux/fs/set_permissions_impl.rs

```diff
@@ -1,49 +1,27 @@
 use super::procfs::set_permissions_through_proc_self_fd;
 use crate::fs::{errors, open, OpenOptions, Permissions};
-use rustix::fs::{fchmod, Mode, OFlags, RawMode};
-use std::os::unix::fs::{OpenOptionsExt, PermissionsExt};
+use rustix::fs::{fchmod, Mode, RawMode};
+use std::os::unix::fs::PermissionsExt;
 use std::path::Path;
-use std::sync::atomic::AtomicBool;
-use std::sync::atomic::Ordering::Relaxed;
 use std::{fs, io};
 
 pub(crate) fn set_permissions_impl(
     start: &fs::File,
     path: &Path,
     perm: Permissions,
 ) -> io::Result<()> {
-    // Record whether we've seen an `EBADF` from an `fchmod` on an `O_PATH`
-    // file descriptor, meaning we're on a Linux that doesn't support it.
-    static FCHMOD_PATH_BADF: AtomicBool = AtomicBool::new(false);
-
     let std_perm = perm.into_std(start)?;
 
-    if !FCHMOD_PATH_BADF.load(Relaxed) {
-        // First try to open the path with `O_PATH`; if that succeeds, it'll give
-        // us a few options. Use `read(true)` even though we don't need `read`
-        // permissions, because Rust's libstd requires an access mode, and Linux
-        // ignores `O_RDONLY` with `O_PATH`.
-        // TODO: Add tests with no-read no-permissions.
-        let opath_result = open(
-            start,
-            path,
-            OpenOptions::new()
-                .read(true)
-                .custom_flags(OFlags::PATH.bits() as i32),
-        );
-
-        // If `O_PATH` worked, try to use `fchmod` on it.
-        if let Ok(file) = opath_result {
-            match set_file_permissions(&file, std_perm.clone()) {
-                Ok(()) => return Ok(()),
-                Err(err) => match rustix::io::Error::from_io_error(&err) {
-                    // If it fails with `EBADF`, `fchmod` didn't like `O_PATH`,
-                    // so proceed to the fallback strategies below.
-                    Some(rustix::io::Error::BADF) => FCHMOD_PATH_BADF.store(true, Relaxed),
-                    _ => return Err(err),
-                },
-            }
-        }
+    // First try using `O_PATH` and doing a chmod on `/proc/self/fd/{}`
+    // (`fchmod` doesn't work on `O_PATH` file descriptors).
+    //
+    // This may fail, due to older Linux versions not supporting `O_PATH`, or
+    // due to procfs being unavailable, but if it does work, go with it, as
+    // `O_PATH` tells Linux that we don't actually need to read or write the
+    // file, which may avoid side effects associated with opening device files.
+    let result = set_permissions_through_proc_self_fd(start, path, std_perm.clone());
```

Comment:
Not that I care much, but in theory we could cache the failure here just like above?

(But eh, anyone running such old kernels has bigger problems IMO)

Reply:
Yeah, we could, but also yeah. It's useful to make things work, to the extent we can, but `O_PATH` goes back to 2.6.39, so it's unlikely this level of optimization is important here.